base/reduce.jl

```diff
@@ -659,6 +659,217 @@ julia> minimum([1,2,3])
 """
 minimum(a) = mapreduce(identity, min, a)
 
+## findmax, findmin, argmax & argmin
+
+"""
+    findmax(f, domain) -> (f(x), x)
+    findmax(f)
+
+Returns a pair of a value in the codomain (outputs of `f`) and the corresponding
+value in the `domain` (inputs to `f`) such that `f(x)` is maximised. If there
+are multiple maximal points, then the first one will be returned.
+
+When `domain` is provided it may be any iterable and must not be empty.
+
+When `domain` is omitted, `f` must have an implicit domain. In particular, if
+`f` is an indexable collection, it is interpreted as a function mapping keys
+(domain) to values (codomain), i.e. `findmax(itr)` returns the maximal element
+of the collection `itr` and its index.
+
+Values are compared with `isless`.
+
+# Examples
+
+```jldoctest
+julia> findmax(identity, 5:9)
+(9, 9)
+
+julia> findmax(-, 1:10)
+(-1, 1)
+
+julia> findmax(cos, 0:π/2:2π)
+(1.0, 0.0)
```

Comment:
I think it'd be nice to do demonstrate "first match wins" behavior. Maybe something like:

```julia
julia> findmax(first, [(1, :a), (2, :b), (2, :c)])
(2, (2, :b))
```

Maybe this was the intention of `cos` example but it is not very obvious to realize this when floating point arithmetic is involved.